src/enumerator/arithmetic_sequence_object.cpp

```diff
@@ -0,0 +1,261 @@
+        auto value = _step->send(env, "*"_s, { IntegerObject::create(i) })->send(env, "+"_s, { _begin });
+
+        // Ensure that we do not yield a number that exceeds `_end`
+        if (!infinite && value->send(env, cmp, { _end })->is_truthy())
+            value = _end;
+
+        func(value);
+    }
+
+    return this;
+}
+
+Value ArithmeticSequenceObject::enum_block(Env *env, Value self, Args args, Block *) {
+    auto yielder = args.at(0);
+    auto enumerator = self->as_enumerator_arithmetic_sequence();
+
+    // each { |value| yielder << value }
+    enumerator->iterate(env, [&yielder, &env](Value value) -> Value {
+        return yielder.send(env, "<<"_s, { value });
+    });
+
+    return enumerator;
+}
+
+bool ArithmeticSequenceObject::eq(Env *env, Value other) {
+    if (!other->is_enumerator_arithmetic_sequence())
+        return false;
+
+    ArithmeticSequenceObject *other_sequence = other->as_enumerator_arithmetic_sequence();
+    return hash(env)->equal(other_sequence->hash(env));
+}
+
+Value ArithmeticSequenceObject::hash(Env *env) {
+    HashBuilder hash_builder {};
+    auto hash_method = "hash"_s;
+
+    auto add = [&hash_builder, &hash_method, env](Value value) {
+        auto hash = value->send(env, hash_method);
+
+        if (hash->is_nil())
+            return;
+
+        auto nat_int = IntegerObject::convert_to_nat_int_t(env, hash);
+        hash_builder.append(nat_int);
+    };
+
+    add(m_begin);
+    add(m_end);
+    add(step());
+
+    if (m_exclude_end)
+        add(TrueObject::the());
+    else
+        add(FalseObject::the());
+
+    return IntegerObject::create(hash_builder.digest());
+}
+
+Value ArithmeticSequenceObject::inspect(Env *env) {
+    switch (m_origin) {
+    case Origin::Range: {
+        auto range_inspect = RangeObject(m_begin, m_end, m_exclude_end).inspect_str(env);
+        auto string = StringObject::format("(({}).step", range_inspect);
+
+        if (has_step()) {
+            string->append_char('(');
+            string->append(m_step);
+            string->append_char(')');
+        }
+
+        string->append_char(')');
+
+        return string;
+    }
+    case Origin::Numeric: {
+        auto string = StringObject::format("({}.step({}", m_begin, m_end);
+
+        if (has_step()) {
+            string->append(", ");
+            string->append(m_step);
+        }
+
+        string->append("))");
+
+        return string;
+    }
+    }
+    return nullptr;
+}
+
+Value ArithmeticSequenceObject::last(Env *env, Value n) {
+    auto _end = maybe_to_f(env, m_end);
+    auto steps = step_count(env);
+
+    if (n) {
+        auto n_as_int = n->to_int(env);
+        Integer count = n_as_int->integer();
+
+        if (count < 0)
+            env->raise("ArgumentError", "negative array size");
+
+        if (n_as_int->integer() > steps)
+            count = steps;
+
+        n_as_int->assert_fixnum(env);
+
+        // NATFIXME: What if we exceed size_t?
```

Comment:
Actually, the real danger is underflowing a size_t. The `count` integer could be negative (unless there is a check for that I'm missing). When casted to a size_t, the number can wrap around.

Let's throw an assert right here to ensure the `count` is positive.

(Also, `size_t` is a 64-bit integer, which presumably is big enough to represent every bye of memory any computer has or will have for a long time. :smile:)

Reply:
There is a check above to verify that count is positive so this should be fine, shouldn't it?

Reply:
OK, very good. Let's just remove the comment then.